### ISSUE TITLE
Introduce a ConfigPhase that is used at build time but is overridable at runtime

### ISSUE THREAD

**Issue by iocanel** [quarkusio/quarkus#33247]
### Description

The last couple of days I've been seen a recurring problem: People using runtime configuration properties in order to influence build time behavior. By people I mean both users and quarkus developers.

At the moment there is no way to have a property that is meant to be overridable at run time, but it's visible at build time. 
Even if the value changes at runtime, the inital intent is really useful for build time tooling (e.g. kubernetes manifest generator etc).

### Use cases
Here as the use cases that come to mind.

#### Kubernetes Manifest Generation

In many cases the manifests we generate try to adapt to the configuration and build items available at build time. We often run into properties we want to use that are runtime properties.

##### Exposing https
When we generate manifests it would help to know if ssl is enabled.

##### Creating Jobs for Initialization Tasks
For extensions like:

- liquibase
- liquibase-mongo
- flyway

We support generating Kubernetes Jobs. However, those extensions may be disabled at runtime. If users have them disabled in the `application.properties` it's should be a strong indication that the manifest generator should ignore them. Or at least this is what most people initially expect (this includes seasoned users and developers).

### Implementation ideas

Create a new ConfigPhase constant like: 

```java
BUILD_TIME_AND_RUN_TIME_OVERRIDABLE(true, true, true, "Build time and run time overridable")
```

**Comment by geoand** [comment]
cc @radcortez @maxandersen @metacosm as we discussed this problem the other day

**Comment by radcortez** [comment]
We need to gather and detail the actual use cases and figure out what to do.

I'm biased on the configuration side, and I admit there may be viable use cases for this feature, but we need to understand how precisely.

The point here is if you use a configuration at build time, you are using it to make some decisions (adding beans, generating some files, etc). If a build configuration is also available at runtime and can be overridden, it may or may not require a reagumentation or some intermediary step that deals with the new configuration value to adjust to the new behaviour defined by the overridden configuration.

Another concern is if we allow such a feature, there is no stopping on how it will be used. It has been long and hard process to educate our users that we have a build configuration that cannot be overridden in exchange for smaller footprints, optimizations, etc. We know that users prefer a dynamic behaviour at runtime, where they can override any configuration and reflect a change in the runtime.

**Comment by radcortez** [comment]
Also /cc @dmlloyd

**Comment by dmlloyd** [comment]
As you all probably know, I've been consistently opposed to such a feature. But it is helpful to have a few use cases to help explain concrete reasons.

Almost every use case I've seen comes down to specifying some set of things at build time to be pre-generated, pre-optimized, or otherwise eagerly set up under the expectation that it will be used by default at run time, while also allowing the run time to change the actual set of things to use - either to make it a subset of the build time set, or simply to specify something different that wasn't initially expected at build time.

However, the problem with this kind of thing starts to become evident when the documentation for the property is written: you generally have to explain what the property does at build time (typically pre-generation for example), and what distinct behavior it has at run time (selection and/or run-time generation for example). The distinct behavior is where I have an issue. If the build time and run time behavior were split into two properties, then it would be natural to describe the build time behavior in the one property and the run time behavior in the other, and it would set the user's expectations clearly.

To give an example, consider the extensions above that generate Kubernetes jobs. Isn't it better for the user to have a property which says "Generate the Kubernetes job" and one which says "Enable or disable the generated Kubernetes job, if any" than one which says "At build time generate the Kubernetes job, and at run time enable or disable it"?  In the former case, if I set the value of this super-property to be a property expression for example, then I have to worry about the differing behavior of that property expression at build time and run time. If the two are completely separated then I don't have this problem since I know exactly when my expression will apply.

Another example which came up was one relating to support for specialized databases. The idea was that at build time the property would give the set of databases which would have pre-generated support, and at run time the property would enable the given set of databases, possibly generating support at run time for the ones not given in the build time set. In this case a very similar issue comes up that the property has a very different meaning at these two times.

In general if I have to read a paragraph of text - even a short one - to understand what the default value for a property is, as a user I'm going to be very annoyed - _especially_ because we _have_ a standard way to record the default value of a property.

However, by making this two distinct properties - say `..frobnibase.supported=a,b,c` at build time and `..frobnibase.enabled` at run time, not only can I document these two distinctly, but I can set the default value of the latter to be a property expression referencing the former, which lets us use our standard default value documentation mechanism and makes it immediately clear what's going on.

I think that one critical question to ask is: "As an extension developer, do I want this to make the user's life easier, or to make my own life easier?". Another is "Is the behavior of this property really explainable/comprehensible to a brand new developer?"

A good idiom to use to think of these kinds of things are to consider having separate properties for `...supported` (build time) vs. `...enabled` (run time). If `supported` is `false` and `enabled` is `true`, you can create a very sensible warning for this situation ("Cannot enable frozbat because it was not configured in the build; see `quarkus.frozbat.supported` for more information" or so).

In conclusion: distinct behaviors should be implemented by distinct properties. Otherwise it's going to be more confusing for the user. Keep the model simple.

**Comment by iocanel** [comment]
Thank you @dmlloyd for the detailed explaination.

While your reponse covered me, there is a tiny issue:

If we have the runtime property defaulting to the build time one, we will end up in cases where a feature flag (run time) will depend on its corresponding kubernetes feature flag (build time).  For exmample in the recent case of liquibase we will end up having `quarkus.liquibase.enabled` depend on `quarkus.liquibase.generate-kubernetes-job`. Which is a bit weird. 

Other than that your response made perfect sense to me. 
Thanks again






> However, by making this two distinct properties - say `..frobnibase.supported=a,b,c` at build time and `..frobnibase.enabled` at run time, not only can I document these two distinctly, but I can set the default value of the latter to be a property expression referencing the former, which lets us use our standard default value documentation mechanism and makes it immediately clear what's going on.

**Comment by radcortez** [comment]
> If we have the runtime property defaulting to the build time one, we will end up in cases where a feature flag (run time) will depend on its corresponding kubernetes feature flag (build time). For exmample in the recent case of liquibase we will end up having `quarkus.liquibase.enabled` depend on `quarkus.liquibase.generate-kubernetes-job`. Which is a bit weird.

Depends... you could record the runtime property `quarkus.liquibase.enabled` to `true` if `quarkus.liquibase.generate-kubernetes-job` is also `true` in a `BuildStep`. You can use `RunTimeConfigurationDefaultBuildItem` to record the property and the value.

**Comment by radcortez** [comment]
I believe we concluded that we are leaving this as is. Please, reopen otherwise.
